### PR TITLE
Add common request and schemas for LLM query interface

### PR DESCRIPTION
**Pull request by harupy**
https://github.com/mlflow/mlflow/assets/17039389/f2a1dfea-4d3a-4786-b07f-39ed82942964

## Related Issues/PRs

 #xxx

## What changes are proposed in this pull request?

Add common request and schemas for LLM query interface.

## How is this patch tested?

- [ ] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/gateway/app.py [hunk 1]
```diff
@@ -48,17 +54,32 @@ async def search_routes():
     return {"routes": ACTIVE_ROUTES}
 
 
-def _add_dynamic_route(route: RouteConfig):
-    async def route_endpoint(request: Request):
-        try:
-            # TODO: handle model-specific route logic by mapping the provider to plugin logic
+async def _chat(_request: chat.Request) -> chat.Response:
+    return chat.Response(**{"candidates": [{"role": "assistant", "content": "hello"}]})
```
Comment (harupy) [inline mlflow/gateway/app.py:58]: A dummy implementation. Real one would 

```python
def create_chat_handler(config: RouteConfig):

    client = get_client(config):
    
    
    async def handler(request: chat.Request) -> chat.Response:
        return client.chat(request)


    return handler
```

File: mlflow/gateway/schemas/completions.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
+from typing import List
+
+from pydantic import BaseModel, Extra
+
+from .common import Metadata
+
+
+class Request(BaseModel, extra=Extra.allow):
```
Comment (harupy) [inline mlflow/gateway/schemas/completions.py:8]: `Request` might be ambiguous. A name like `RequestBody` or `RequestPayload` might be better.

File: mlflow/gateway/schemas/completions.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    prompt: str
+
+
+class Candidate(BaseModel, extra=Extra.allow):
+    text: str
+    metadata: Metadata
+
+
+class Response(BaseModel, extra=Extra.allow):
```
Comment (harupy) [inline mlflow/gateway/schemas/completions.py:17]: `ResponsePayload` might be better.

File: mlflow/gateway/schemas/chat.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+
+class Candidate(BaseModel):
+    message: Message
+    metadata: Metadata
+
+
+class Response(BaseModel, extra=Extra.allow):
+    candidates: List[Candidate]
+    metadata: Metadata
```
Comment (harupy) [inline mlflow/gateway/schemas/chat.py:24]: @dbczumar Is metadata an optional field?

File: mlflow/gateway/schemas/common.py [hunk 5]
```diff
@@ -0,0 +1,3 @@
+from typing import Dict, Any
+
+Metadata = Dict[str, Any]
```
Comment (harupy) [inline mlflow/gateway/schemas/common.py:3]: Each route has a differen metadata. This needs to be fixed.

File: mlflow/gateway/schemas/completions.py (hunk 2 above)
Comment (harupy) [reply mlflow/gateway/schemas/completions.py:8]: `RequestPayload` acutally sounds good.

File: mlflow/gateway/app.py [hunk 6]
```diff
@@ -53,17 +53,37 @@ async def search_routes():
...
+        **{"embeddings": [random.randint(0, length - 1) / length for _ in range(10)]}
+    )
+
+
+def _route_type_to_endpoint(route_type: RouteType):
+    if route_type == RouteType.Chat:
+        return _chat
+    elif route_type == RouteType.Completions:
+        return _completions
+    # elif route_type == RouteType.Embeddings:
```
Comment (BenWilson2) [inline mlflow/gateway/app.py:78]: did we want this to be commented out?

File: mlflow/gateway/schemas/chat.py (hunk 4 above)
Comment (BenWilson2) [reply mlflow/gateway/schemas/chat.py:34]: I think for these providers that we're going to have to allow extras for probably all fields in the responses received. 
Anthropic's API docs: https://docs.anthropic.com/claude/reference/versioning 
Seems like they're using metadata (at the moment) for just a user uuid, but that could be added to without bumping the endpoint version according to their docs.

Comment (harupy) [reply mlflow/gateway/schemas/chat.py:34]: Got it, will update.

File: mlflow/gateway/app.py (hunk 6 above)
Comment (harupy) [reply mlflow/gateway/app.py:78]: I commented this out because we don't have `RouteType.Embeddings` yet

File: tests/gateway/schemas/test_chat.py
Comment (harupy) [inline tests/gateway/schemas/test_chat.py:1]: I'll add more test cases for the input payload.

File: mlflow/gateway/schemas/chat.py (hunk 4 above)
Comment (harupy) [reply mlflow/gateway/schemas/chat.py:42]: Done!

File: examples/gateway/config.yaml [hunk 7]
```diff
@@ -1,6 +1,6 @@
 routes:
-  - name: gpt-3.5-turbo-completions
-    type: llm/v1/completions
+  - name: gpt-3.5-chat
+    type: llm/v1/chat
```
Comment (harupy) [inline examples/gateway/config.yaml:3]: Just wanted to shorten these for quicker typing.

File: mlflow/gateway/app.py (hunk 6 above)
Comment (dbczumar) [reply mlflow/gateway/app.py:116]: Can we uncomment and add it?

**Review by dbczumar — APPROVED** [review]
LGTM once comments addressed. Thanks @harupy !

File: mlflow/gateway/app.py (hunk 6 above)
Comment (harupy) [reply mlflow/gateway/app.py:116]: Uncommented, #8711 will be merged soon.

**Pull request merged**